### DISCUSSION TITLE
Looping (TG1WDT_SYS_RESET) error using WiFi.h

### DISCUSSION THREAD

**Discussion by BBwally** [community/community#139342] [Programming Help]
### Body

Hi - and forgive me if I have failed to find an earlier solution to this issue. I had an application running nicely on a LilyGo T-Display ESP32 module - and suddenly the WiFi connection no longer works. It was previously working quite happily with TTGO LoRa32-OLED board definition. I have distilled the minimum code needed to reproduce the error as the following:

#include <WiFi.h>

const char* SSID="MyNetwork";
const char* PW="MyPassword";

void setup() {
  char * pSSID;

  //initialise serial port
  Serial.begin (115200);
  while (!Serial) {delay (100);}
  Serial.println ("Confirming Serial");
  delay (5000);

  //try and connect to WiFi
  pSSID=new char [strlen (SSID)+1];
  strcpy (pSSID, SSID);
  Serial.println ("About to attempt WiFi connection");
  WiFi.begin (pSSID, PW);
  delay (5000);
  Serial.println ("Just issued WiFi.begin call");
  while (WiFi.status() != WL_CONNECTED) {
    Serial.println ("Waiting");
    delay (1000);
  }
  Serial.println ("Connected!");
  delete [] pSSID;
}

void loop() {
  Serial.println ("Whew! Out of setup");
  delay (10000);
}

This code commences looping immediately after the "About to attempt WiFi connection" output. The full reboot message is as follows:

14:51:43.106 -> Confirming Serial
14:51:48.138 -> About to attempt WiFi connection
14:51:49.075 -> ets Jun  8 2016 00:22:57
14:51:49.075 -> 
14:51:49.075 -> rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
14:51:49.075 -> configsip: 0, SPIWP:0xee
14:51:49.075 -> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
14:51:49.075 -> mode:DIO, clock div:1
14:51:49.075 -> load:0x3fff0030,len:4604
14:51:49.075 -> ho 0 tail 12 room 4
14:51:49.075 -> load:0x40078000,len:15488
14:51:49.075 -> load:0x40080400,len:4
14:51:49.075 -> load:0x40080404,len:3180
14:51:49.075 -> entry 0x400805b8

If anyone is able to help me with some tips as to what may be going wrong, I will be eternally grateful!

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by jgschmitz** [comment]
Hi @BBwally 
The ESP32 has a watchdog timer to prevent tasks from hanging. The delay(5000) calls may be too long without giving control back to the system, which can trigger the watchdog reset. Try replacing the long delay(5000) in setup() with a shorter delay and call delay() in smaller increments or use yield().
```
void setup() {
    char * pSSID;

    Serial.begin(115200);
    while (!Serial) { delay(100); }
    Serial.println("Confirming Serial");

    for (int i = 0; i < 50; i++) {
        delay(100);  // Replaces delay(5000)
        yield();     // Allows ESP32 to handle system tasks
    }

    pSSID = new char[strlen(SSID) + 1];
    strcpy(pSSID, SSID);

    Serial.println("About to attempt WiFi connection");
    WiFi.begin(pSSID, PW);
    delete[] pSSID;

    // Check WiFi connection status in shorter intervals and allow yielding
    while (WiFi.status() != WL_CONNECTED) {
        Serial.println("Waiting");
        delay(500);  // Reduces the delay time
        yield();
    }

    Serial.println("Connected!");
}
```

Also - ESP32 may get stuck in the WiFi.begin() call, especially if it is not able to connect to the network. You can add a timeout to avoid this:
```
void setup() {
    Serial.begin(115200);
    while (!Serial) { delay(100); }
    Serial.println("Confirming Serial");

    WiFi.begin(SSID, PW);
    Serial.println("About to attempt WiFi connection");

    unsigned long startAttemptTime = millis();

    // Retry for 10 seconds
    while (WiFi.status() != WL_CONNECTED && millis() - startAttemptTime < 10000) {
        Serial.println("Waiting for connection...");
        delay(500);
        yield();
    }

    if (WiFi.status() == WL_CONNECTED) {
        Serial.println("Connected!");
    } else {
        Serial.println("Failed to connect to WiFi");
    }
}
```

Let me know if either of those two things are helpful - cheers 🍻

**Reply by BBwally** [reply]
Thanks Jeff for these suggestions. I incorporated all of the suggested changes, but unfortunately no luck. The code consistently crashes approximately 900ms after the WiFi.begin call. In earlier working versions, I suspect I was using a different library as I was able to call WiFi.begin with both SSID and PW as const char* parameters. I did accept an IDE suggestion to update some of my libraries, and that's when the problems began - SSID was no longer accepted as a const char* parameter - and hence the couple of lines code to transcribe it into a char* parameter. Do you have any suggestions about an alternative WiFi library to try?

**Comment by Chanduporalla** [comment]
It seems that your ESP32 module is encountering a watchdog reset (TG1WDT_SYS_RESET), which can happen due to long delays or an issue with the task not being handled properly by the system. Here are a few suggestions to troubleshoot and potentially resolve the issue:

    Increase the Delay after WiFi.begin(): Sometimes, ESP32 boards need a bit more time for WiFi to initialize. Increase the delay after calling WiFi.begin() to 10,000 ms (10 seconds) to see if it helps.

    cpp

WiFi.begin(pSSID, PW);
delay(10000);

Watchdog Timer: The ESP32 has a watchdog timer that resets the CPU if the code takes too long without allowing the system to manage background tasks. Try adding yield() or delay(1) inside your loop to prevent the watchdog reset.

cpp

while (WiFi.status() != WL_CONNECTED) {
    Serial.println("Waiting for connection...");
    delay(1000);
    yield();  // Allow the system to handle background tasks
}

Check Power Supply: Ensure that your ESP32 is getting sufficient power. If you're powering the module via USB, try using a more stable 5V power source as insufficient power can sometimes cause reboot loops or resets during WiFi operations.

WiFi.begin() without parameters: You can try calling WiFi.begin() without parameters if the credentials have been previously stored in the flash memory. Sometimes this can work around WiFi issues:

cpp

    WiFi.begin();

    Flash Erase: If nothing seems to work, a full erase of the flash and then re-flashing the code might help clear any corrupted settings or memory.

If none of the above solves the problem, it may be helpful to post this on the GitHub ESP32 community or the LilyGo T-Display forum, as hardware-specific issues can sometimes arise from board definitions or firmware.

**Reply by BBwally** [reply]
Thanks Chanduporalla. I've now tried three suggestions (independent 18w PSU, full flash erase and including the yield() function in a loop with a shortened delay) - no luck there. I'll need to research how to store the SSID and PW in flash and give that a try.

**Comment by BBwally** [comment]
Thanks to the two folk who responded. The problem turned out to be corrupted board definitions and was remedied by refreshing the ESP32 profiles - so all sorted now. However, I've hardened the code a notch further by implementing some of the "good practice" suggestions that were offered ... like including the yield() function in appropriate places.

**Comment by Chanduporalla** [comment]
Thank you for the update! I'm glad to hear that refreshing the ESP32 profiles resolved the issue with the corrupted board definitions. It's great that you've also taken the time to implement the good practice suggestions. Including the yield() function is definitely a wise move for improving code stability. If you encounter any more challenges or have further questions.
